man/lpr.1
.ds VE LPRng-3.2.1
.TH LPR 1 \*(VE "LPRng"
.ig
Thu Jul 20 07:19:29 PDT 1995 Patrick Powell
..
.SH NAME
lpr \- off line print
.SH SYNOPSIS
.B lpr
[
.B \-A
]
[
.B \-b,l
]
[
.BI \-C " class"
]
[
.BI \-D " debugopt"
]
.ti +5n
[
.BI \-F " filterformat"
]
[
.B \-h
]
[
.BI \-i " indentcols"
]
.ti +5n
[
.B \-k
]
[
.BI \-J " job"
]
[
.BI \-K,# " copies"
]
.ti +5n
[
.BI \-m " mailTo"
]
[
.BI \-P " printer"
]
[
.B \-Q
]
[
.B \-r
]
.ti +5n
[
.BI \-R " remoteAccount"
]
[
.B \-s
]
[
.BI \-T " title"
]
[
.BI \-U " user"
]
.ti +5n
[
.B \-V
]
[
.BI \-w " width"
]
[
.BI \-Z " zoptions"
]
.ti +5n
[
.IR filename " \|.\|.\|."
]
.SH DESCRIPTION
.B Lpr
uses a spooling daemon to print the named files when facilities
become available.  If no names appear, the standard input is assumed.
.SH OPTIONS
.TP 5
.B \-A
Use the authenticated transfer facilities if they are available.
.TP 5
.BI \-C " class"
Specify the job classification for use on the burst page and to
set the priority.
Priorities range from A (highest) to Z (lowest);
the default priority is A.
For example,
.br
.ti +0.5i
lpr \-C B foo.c
.br
sets the priority/class to B and the file foo.c is printed.
.TP 5
\fB\-D\fP\ \fIdebugoptions\fP
Debugging is controlled using the
.B \-D
option. This accepts a comma-separated list of debugging
settings. These settings take one of two forms: 
.BI facility =value, 
or
.B value 
to set an overall default value.
.TP 5
\fB\-F\fI filterformat\fR
Filter or format specification.
By default,
input is assumed to a standard text file and the
.I f
format is used;
the output device is assmed to be a simple line printer.
Therer are other formats available,
listed below.
Not all formats may be available on all printers;
see
.IR printcap (5)
for details.
Formats are single lower case letters;
the following are the valid arguments for
.B \-F
together with the assumed type of data.
For compatibility with previous versions of
.BR lpr ,
the format types can be used as options themselves
(i.e. by omitting the
.BR F )
except where noted below,
a warning may be issued in such cases.
.RS 5
.TP 5
.B "\-b, \-l"
either of these flags specifies a
binary or literal file,
and no processing is to be done.
Printed using the
.I f
format filter,
with processing suppressed.
.TP 5
.B \-c
data produced by
.IR cifplot (l).
.TP 5
.B \-d
output from
.IR tex (l)
(DVI format from Stanford).
.TP 5
.B \-g
standard plot data as produced by the
.IR plot (3X)
routines (see also
.IR plot (1G)
for the filters used by the printer spooler).
.TP 5
.B \-n
output from (device independent)
.IR troff .
.TP 5
.B \-p
First use
.IR pr (1)
to format the files,
then print using 
.I f
format.
.TP 5
.B \-r
text in which the first character of each line is interpreted as a
standard FORTRAN carriage control character.
The effect of this format used to be obtained with the
.B \-f
option.
.TP 5
.B \-t
output from
.IR troff (1)
(originally cat phototypesetter commands,
but now we assume the same as the
.B n
format).
.TP 5
.B \-v
a raster image for devices like the Benson Varian.
.RE
.TP 5
\fB\-h\fP
No banner or header for this job.
.TP 5
.BI \-i " indentcols"
Indent input by indentcols.
Note that this option is not supported on all printers.
.TP 5
.BI -J " jobname"
Specify the job name to print on the burst page;
defaults to the name of the first file in job or
.I "(STDIN)"
if input is from a pipe.
.TP 5
.B \-k
When used as a filter, lpr normally creates a temporary file
for the input read from stdin before sending it to the remote
printer.
The
.B \-k
causes the job to be sent directly to the server.
If you kill the job in the middle of creation,
then the partly transferred file will get printed.
This option may not work with very large jobs or non-LPRng spoolers.
.TP 5
\fB\-K\fP\fIcopies\fP,\0\fB\-#\fP\fIcopies\fP
Specify the number of copies of each file to be printed.
.TP 5
.BI \-m " mailTo"
Send mail upon unsuccessful completion to user
.IR mailTo .
.TP 5
.BI \-P " printer"
Output to the specific printer;
the default is (in order of priority)
the explictly specified -P value,
the environment variable PRINTER,
the first entry in the printcap information,
and the default_printer entry from the
configuration file,
and finally the hard coded (site dependent) default_printer
value.
.TP 5
.B \-Q
Put the name of the spool queue into the job file.
This information can then be used by the spooling software
to control output format.
.TP 5
.BI \-R " remoteAccount"
Specify accounting information to be used by a remote system that prints
your output.
.sp
This parameter
can be used to specify a billing code to be charged for the
printing.
.TP 5
.B \-s
This flag is included for compatibility with other versions of lpr.
In these versions it will create a symbolic link to the files to be
printed.  
.I Lpr
now sends files directly to the server and it is irrelevant.
.TP 5
.BI \-T "  title"
Specify the title used by
.IR pr (1);
defaults to the file name.
.TP 5
\fB\-U\fP\ \fIusername\fP
The
.B \-U
option is used to specify a user name
for the job.
This is obviously a security loophole,
but it is present for compatibility reasons.
See
.B "Authenticated Transfers"
below.
.TP 5
.B \-V
Verbose mode. Additional -V flags increase verbosity.
Use debug flags for extreme verbosity.
.TP 5
.BI \-w " width"
Specify the page width for printing the job.
.TP 5
.BI -Z " zoptions"
This option is used to pass options to the print spooler,
and is used when additional or specialized information
must be provided to the spooler.
.TP 5
.BI \-1,2,3,4 "fontname"
Specify a font to be mounted on font position \fIi\fR
for TROFF printing (Obsolete).
.SH FILENAMES
.PP
By default,
if no filenames are specified
.I lpr
will read stdin and print it.
A filename of the form
.I -
(single minus sign)
will take input from STDIN (file descriptor 0).
.SH "AUTHENTICATED TRANSFERS"
.PP
The original LPR network protocol defined in RFC1179 did not
provide for user to server authentication.
This is now supported by LPRng.
See the LPRng support documentation for details on its operation
and support.
.SH COMPATIBILITY 
.PP
The LPRng version of lpr
attempts to be functionaly compatible with common implementations of lpr.
However,
there are some commands and functionality that are deliberately missing.
.IP "\fB\-s\fP\ Symbolic Links" 5
(Berkeley LPR) This option specified that a symbolic link
to the original data file
rather than a copy of the data file was to be used when spooling jobs.
This opens up a variety of security problems,
as well as being ineffective when printing to a remote host.
.IP "\fB\-r\fP\ Remove File" 5
This option specified that data files were to be removed after
the print job was completed.
This opens up a variety of security problems,
especially in a NFS based system.
.SH FILES
.PP
The files used by LPRng are set by values in the
printer configuration file.
The following are a commonly used set of default values.
.nf
.ta \w'/var/spool/lpd/printcap.<hostname>           'u
/etc/lpd.conf		LPRng configuration file
/etc/printcap		printer description file
/etc/lpd.perms	printer permissions
/var/spool/printer*		spool directories
/var/spool/printer*/printer	lock file for queue control
/var/spool/printer*/control.printer	queue control
/var/spool/printer*/active.printer	active job
/var/spool/printer*/log.printer	log file
.fi
.SH "SEE ALSO"
lpd.conf(5),
lpc(8),
lpd(8),
lpr(1),
lpq(1),
lprm(1),
printcap(5),
lpd.perms(5),
pr(1).
.SH DIAGNOSTICS
.nf
Most of the diagnostics are self explanatory.
If you are puzzled over the exact cause of failure,
set the debugging level on (-D5) and run again.
The debugging information will 
help you to pinpoint the exact cause of failure.
.fi
.SH "HISTORY"
.LP
LPRng is a enhanced printer spooler system
with functionality similar to the Berkeley LPR software.
In 1988 Patrick Powell released
the PLP (Public Line Printer) software,
which went through several evolutions.
Justin Mason (jmason@iona.ie)
generated PLP4.0 from several older releases of PLP.
In 1992 Patrick Powell
release LPRng,
a completely redesigned and newly written version of the software.
.LP
The LPRng mailing list is plp@iona.ie;
subscribe by sending mail to plp-request@iona.ie with
the word subscribe in the body.
The software is available from ftp://iona.ie/pub/LPRng.
.LP
LPRng is distributed under the GNU software license for non-commercial
use,
the Artistic License for limited commercial use. 
Commerical support and licensing is available through
Patrick Powell <papowell@sdsu.edu>.


NEWS
---------------------------------------------------------------------
Version LPRng-3.8.A - not yet released
---------------------------------------------------------------------

This is the first release not done by Patrick Powell but
by the new team to maintain LPRng. The project is hosted
on sourceforge, take a look at
http://sourceforge.net/projects/lprng/
and join us on the new mailing-list at
http://lists.sourceforge.net/mailman/listinfo/lprng-devel

The biggest change is the switch to using automake to generate
the Makefile.im and cleaning the autoconf rules a bit. Most
things should work just as before, with the following exceptions:

- make install no longer tries to remove previous installed
  printing systems and to modify your system to run LPRng, you
  will have to do this yourself. Example files with descriptions
  can be found in the conf/ subdirectory of the source.
- binaries are currently not made suid even if requested,
  that might be reimplemented in the future.
- when configuring with --enable-nls less time is invested in
  finding a gettext implementation. You might need to give it
  some more flags when using a gettext not in your libc.

There is a new configure option --disable-remote to change the default
to not listen via tcp and not accepting requests from remote hosts.

Otherwise a lot of cleanup has happened. Many compiler warnings
quieted, prototypes added, format strings synced with format
of arguments, some code and manpages with strange license situation
replaced, local functions declared static, moving some code around
to not have to link all lpd code in all other binaries...

---------------------------------------------------------------------
Changes in versions until LPRng-3.8.28 can
be found in the file CHANGES.
If that file did not come with your LPRng, you should be able to download
it from: http://lprng.cvs.sourceforge.net/*checkout*/lprng/lprng/CHANGES


src/LPR/lpr.c
/***************************************************************************
 * LPRng - An Extended Print Spooler System
 *
 * Copyright 1988-1997, Patrick Powell, San Diego, CA
 *     papowell@sdsu.edu
 * See LICENSE for conditions of use.
 *
 ***************************************************************************
 * MODULE: lpr.c
 * PURPOSE:
 **************************************************************************/

/***************************************************************************
 * SYNOPSIS
 *      lpr [ -PPrinter ] [-Q] [ -Knum ] [ -C class ] [ -J job ] [
 *      -RremoteAccount ] [ -m[mailTo] ] [ -T title ] [-i[numcols]]
 *      [ -1234 font ] [ -wnum ] [ -Zzoptions ] [ -Uuser ] [ -HHost
 *      ] [ -Ffilter ] [ -bhrs ] [ -Dn ] [ -X ] [ filename ...  ]
 * DESCRIPTION
 *      Lpr uses a spooling server to print the named files when
 *      facilities become available.  If no Names appear, the stan-
 *      dard input is assumed.
 *      -C class or priority (A - Z)
 *      -D[n] debug level
 *      -F?  Filter or format specification
 *      -HHost Used by root process to specify a Host
 *      -J jobname  Specify the job name to print on the burst page
 *      -Q include the queue name in the control file
 *      -PPrinter  Output to the specified Printer
 *      -R remoteAccount
 *      -T title specify the title used by pr(1);
 *      -Uuser Used by root process to specify a user
 *      -Z output printer filter options
 *      -b The files are assumed to contain binary data
 *      -c data produced by cifplot(l).
 *      -d output from tex(l) (DVI format from Stanford).
 *      -f same as -Fr, FORTAN carriage control
 *      -g standard plot data as produced by the plot(3X) routines
 *      -i[numcols] Cause the output to be indented
 *      -l (literal) text with control characters to be printed
 *      -n output from DITROFF
 *      -m[mailTo] Send mail upon completion
 *      -t output from troff(1)
 *      -v a raster image for devices like the Benson Varian.
 *      -wwidth  specify the page width for pr.
 *      -#num number of copies of each file to be printed.
 ****************************************************************************
 * Implementation:
 * 	Each time lpr is invoked it creates a "job" entry in the appropriate
 * spool directory.  Each job consists of a control file and zero or more
 * data files.  The control file contains lines that specify the job
 * parameters, such as job Name, etc., and the Names of the data files.
 *      Control file format
 *      First character in the line flags kind of entry, remainder of line is
 *          the arguemnt.
 *
 * The control file has the flag lines in the following order:
 *
 * HPJCNMDTL <- Standard SUN/BSD order
 *   ABEFGIKOQRSVWXYZ <- PLP additional ones
 * The following are one per data file:
 *   N      -  file 'name'
 *   a-z    -  data file format and file
 *   U      -  unlink
 *
 * Flag Lines:
 *		C -- "class name" on banner page
 *		H -- "Hostname" of machine where lpr was done (Fully Qualified Domain)
 *		I -- "indent" amount to indent output   - Number
 *		J -- "job Name" on banner page          - String
 *		L -- "literal" user's Name to print on banner - STring
 *		M -- "mail" to user when done printing
 *		N -- "Name" of file (used by lpq)
 *		P -- "Person" user's login Name
 *		Q -- "Queue" name
 *		R -- account id  for charging
 *		U -- "unlink" Name of file to remove after we print it
 *		W -- "width" page width for PR
 *		T -- Title for use with 'p' formatting
 *		X -- "header" header title for PR
 *		Z -- xtra options to filters
 *
 *	Lower case letters are formats, and are together with the Name
 *      of the file to print
 *		f -- "file Name" Name of text file to print
 *		l -- "file Name" text file with control chars
 *		p -- "file Name" text file to print with pr(1)
 *		t -- "file Name" troff(1) file to print
 *		n -- "file Name" ditroff(1) file to print
 *		d -- "file Name" dvi file to print
 *		g -- "file Name" plot(1G) file to print
 *		v -- "file Name" plain raster file to print
 *		c -- "file Name" cifplot file to print
 ***************************************************************************/
/*
Additional Implementation Notes

Wed Apr  5 05:10:45 PDT 1995 Patrick Powell
The control file is the most difficult part of making a portable front end,
as it must be compatible with several existing systems.  There are historical
restrictions on the order and contents of the various fields.

A - Job Id (LPRng)
  Identifier for the job.  This should be unique for each job
  that is generated.
H - Hostname     (RFC1179 - 31 or fewer octets)
  The Hostname field is used to identify the host originating the job.
  However,  RFC1179 does not specify if this is a 'short' or 'fully
  qualified domain name'.  In fact,  on some systems it is impossible to
  provide a fully qualified domain name.  So we have to do the following:
  1. If the control file format string has an S flag, we use the short
     name (overrides all other considerations).
  2. We get the fully qualified domain name and use this.
  3. If we cannot get the fully qualified domain name,  we get the
     short host name.
  4. If we cannot get the short host name,  we use IP address with IP prefixed
	 i.e. - IP130.191.163.56

C - Class Name     (RFC1179 - 31 or fewer octets)
  The Class can be specified on the command line by the -C option
  in almost all LPR implementions,  and is placed in the C flag line.
  If no job class is specified,  the SUN and older Berkeley implementations
  put in the originating host name; PLP puts in the job priority.
  NOTE: some implementations INSIST on the C field being present and non-empty.

J - Job name     (RFC1179 - 99 or fewer octets)
  The Job name can be specified on the command line with the -J option;
  if not specified,  defaults to the list of files to be printed.
  If no file,  the name is '(stdin)'
  NOTE: some implementations INSIST on the J field being present and non-empty.

L - Banner User Name (why L?) (RFC1179 - silent on length)
  If this field is specified,  a banner will be printed.  Note that
  in some LPD implementations that L field triggers banner printing;
  the J field must appear before the L field.  The L field value
  should be 
  1. Set by the -U option (overrides)
  2. The 'P' field (user login name)

	Note: Class and Jobname must precede in control file
	which means that if you specify -U, you must force J and C entries
	in control file.

T - Title for use with 'p' formatting
  If 'p' format (print with 'pr') is specified,  then this is the title 
  to be supplied to 'pr'.

P - "Person" or user's login name
  This is the login name of the user.  It is obtained by trying the
  following in order:
	1. user name entry returned by 'getpwuid( user ID )' 
	2. On PCs or others without password files,  $USER environment variable
	3. On PCs or others without password files,  "nobody"

M - mail on completetion
  This is set by the -m option,  or the LPR tries to infer the user
  and host.  Note that this may not work correctly in systems where
  you work on workstations but have your mail delivered to a central
  site.  The following is used:
  1. -m option value (if any)
  2. Person@fully qualified domainname (note: this may not be the same as
      the H or 'host' information, as the H value may be the short host
      name.

Data Files

Data files are specified in two ways: as a list of names on the command
line OR as the output from another program piped to STDIN. We use an
array of struct datafile{} records to record information about the
data files.

We can use two possible methods for sending the data files.
Method 1:
Pass1: we open each file, get file descriptor fd, stat the fd, close fd.
Pass2: reopen each file, get file descriptor fd, stat the fd,
   compare old and new LENGTH for change, copy the file

Method 2:
Pass1: we open each file, get file descriptor fd, stat the fd
Pass2: restat the file, check for length change, copy the file

Method 1 requires a larger number of open and close operations;
Method 2 requires a larger number of open file descriptors being available;

We will use Method 1, even though it has a slightly higher overhead.

STDIN Spooling:
When we need to create a spool file for STDIN input,  we will use
the following method:
1. Get the temporary directory name:
	$(TMP) environment variable,
	$(TEMP) environment variable,
	"/tmp"
2. Open a file in the temporary directory with a name of the form
	lpr_$PID
3. Write the contents of stdin to this file.
4. After transferring the file to the remote host,  remove the file.

Control File Name

Each control file has the format cfXNNNhost, where NNN is a job number
identifier.  On UNIX systems, NNN is PID%1000, ie.- the process ID
of the LPR process modulo 1000.  If you are using a PC or some other
system that does not have PID's,  then you need another method:
the current time of day % 1000 is pretty good,  but may run
into difficulties if you are printing files quickly. A difficulty
with the PC is that there is no simple way to do a 'lock' operation
to prevent multiple updates.  The solution that appears to
be fairly reasonable is to choose a random number based on the
current following information:
1. time of day to the finest resolution.
2. names of files to be printed.
3. generate a random number using this as a seed.

Note that there is still a 1/1000 chance of getting a duplicate job name;
if you rapidly print the same files multiple numbers of times you most
certainly will get a duplicate.

*/

static char *const _id =
"$Id: lpr.c,v 3.9 1997/03/24 00:45:58 papowell Exp papowell $";

#include "lp.h"
#include "dump.h"
#include "initialize.h"
#include "killchild.h"
#include "pathname.h"
#include "sendjob.h"
#include "setuid.h"
#include "printcap.h"
/**** ENDINCLUDE ****/

char LPR_optstr[]    /* LPR options */
 = "1:2:3:4:#:AC:D:F:J:K:NP:QR:T:U:VZ:bcdfghi:lkm:nprstvw:" ;

/***************************************************************************
 * main()
 * - top level of LPR Lite.  This is a cannonical method of handling
 *   input.  Note that we assume that the LPD daemon will handle all
 *   of the dirty work associated with formatting, printing, etc.
 * 
 * 1. get the debug level from command line arguments
 * 2. set signal handlers for cleanup
 * 3. get the Host computer Name and user Name
 * 4. scan command line arguments
 * 5. check command line arguments for consistency
 * 5. if we are spooling from stdin, copy stdin to a file.
 * 6. if we have a list of files,  check each for access
 * 7. create a control file
 * 8. send control file to server
 *
 ****************************************************************************/

int main(int argc, char *argv[], char *envp[])
{
	off_t job_size;
	struct dpathname dpath;
	struct printcap_entry *printcap_entry = 0;


	/*
	 * set up the uid
	 */
	Errorcode = 1;
	Interactive = 1;
	Initialize();

	/* set signal handlers */
	(void) plp_signal (SIGHUP, cleanup);
	(void) plp_signal (SIGINT, cleanup);
	(void) plp_signal (SIGQUIT, cleanup);
	(void) plp_signal (SIGTERM, cleanup);


	/* scan the argument list for a 'Debug' value */
	Get_debug_parm( argc, argv, LPR_optstr, debug_vars );

	Setup_configuration();

	/* scan the input arguments, setting up values */
	Get_parms(argc, argv);      /* scan input args */

	/* Note: we may need the open connection to the remote printer
		to get our IP address if it is not available */

	Check_parms( &printcap_entry );

	if(DEBUGL4 ) dump_parms("LPR Vars after checking parms",Lpr_parms);

	/*
	 * Fix the rest of the control File
	 */
	job_size = Make_job( Cfp_static );
	if( job_size == 0 ){
		Errorcode = 1;
		Diemsg (_("nothing to print"));
	}

	/* Send job to the LPD server for the printer */

	Init_path( &dpath, (char *)0 );

	Send_job( RemotePrinter?RemotePrinter:Printer, RemoteHost, Cfp_static,
		&dpath, 1, Connect_timeout,
		Connect_interval, Send_timeout, printcap_entry );

	/* the dreaded -r (remove files) option */
	if( Removefiles ){
		int i;

		/* eliminate any possible game playing */
		To_user();
		for( i = 0; i < Filecount; ++i ){
			if( unlink( Files[i] ) == -1 ){
				Warnmsg(_("Error unlinking '%s' - %s"),
					Files[i], Errormsg( errno ) );

			}
		}
	}
	Errorcode = 0;
	cleanup(0);
	return(0);
}
